src/fiber/fiber.ml

```diff
@@ -206,15 +197,40 @@ end = struct
   let reraise_all exns =
     let backup = !current in
     add_refs (List.length exns - 1);
-    List.iter exns ~f:(forward_exn_with_bt backup)
-
-  let new_run f =
+    List.iter exns ~f:(forward_exn_with_backtrace backup)
+
+  let rec run_jobs jobs =
+    try
+      while not (Queue.is_empty jobs) do
+        let (Job ({ run; ctx }, x)) = Queue.pop_exn jobs in
+        current := ctx;
+        run x
```

Comment:
Isn't it simpler to just surround the `run x` with `try .. with`?  You won't even need to make `run_jobs` recursive then.

Reply:
It is, but on the other hand it means setting up an exception handler for each iteration of the inner loop. The current code is more efficient by only setting it up once (or rather once + once per exception raised).